### DISCUSSION TITLE
Support minimumReleaseAge with Python packages on Artifactory

### DISCUSSION THREAD

**Discussion by intgr** [renovatebot/renovate#24829] [Suggest an Idea]
### What would you like Renovate to be able to do?

I'm self-hosting Renovate and I notice that `minimumReleaseAge` does not apply to Python packages.

I also have a local mirror of PyPI via Artifactory, using Poetry with custom `tool.poetry.source` configuration.

It's not explicitly documented which datasources support this and which don't. The public PyPI instance seems to know release dates of packages. Is this something that's just not implemented yet, or could this be an issue due to my PyPI mirror?

The `stabilityDays` documentation only mentions:

> Some datasources do not provide a release timestamp (in which case this feature is not compatible), and other datasources may provide a release timestamp but it's not supported by Renovate (in which case a feature request needs to be implemented).

<details>
<summary>(Not relevant)</summary>

### If you have any ideas on how this should be implemented, please tell us here.

The public PyPI instance seems to return release dates in its JSON API, although not in a very convenient format: https://pypi.org/pypi/django/4.1.5/json

```json
{
    "info": {
        ...
    },
    "urls": [
        {
            "filename": "Django-4.1.5-py3-none-any.whl",
            "upload_time": "2023-01-02T07:34:45",
            "upload_time_iso_8601": "2023-01-02T07:34:45.541744Z",
            ...
        },
        {
            "filename": "Django-4.1.5.tar.gz",
            "upload_time": "2023-01-02T07:34:49",
            "upload_time_iso_8601": "2023-01-02T07:34:49.562776Z",
            ...
        }
    ]
}
```

</details>

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
It's likely because locally Renovate is using the "simple" PyPI interface, which does not contain timestamps. Take a look at the http statistics at the end of a repository run to see which URLs of your local pypi server are being accessed.

**Comment by intgr** [comment]
Right you are, [PyPI Simple Repository API](https://peps.python.org/pep-0503/) does not provide release dates. Are there any work-arounds?

**Comment by rarkins** [comment]
Does your server provide a non simple API with time stamps?

**Comment by intgr** [comment]
Ah sorry, I forgot to reply. I'm using Artifactory, it doesn't support PyPI-like JSON API yet: https://jfrog.atlassian.net/browse/RTFACT-16101

There is, however, an Artifactory-specific API. Not sure if anyone is interested in impementing that: https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API

For my use case, it would also be fine if Renovate used PyPI directly for package metadata.

----

PS: There's even a comment in that ticket regarding Renovate:

> This feature would also be super helpful for us, as [renovatebot](https://github.com/renovatebot/renovate) relies on the JSON API to grab release notes for PyPI updates. Our artifactory mirror only supports the /simple API which does not list a corresponding Github repository, and thus, Renovate cannot find the changelog.

FWIW changelog fetching works in my use case.

**Comment by mrgavinconway** [comment]
Any movement on this? We've got artifactory and would love to use Renovate but are getting a "tool releases not found" error when it queries it and a HTTP 406 error message.

**Comment by gracevivi523** [comment]
I am in a exact same situation, there is a new ticket opened with JFrog Artifactory https://jfrog.atlassian.net/browse/RTFACT-30744  ~ 1month ago, but so far no any updates, the ticket is not even assigned. 

Is there any other alternative solution to check the python package release dates and tell Renovate not update if the release date is too new?  I wrote a custom script to get the release date by directly querying public PyPI JSON API which worked perfectly, but I couldn't find a elegant way to hook it up to Renovate, I tried using `postUpgradeTasks` but I don't think this fits the scenario, as it's being executed after the updates. While in our scenario, we wanna renovate skip the updates if the package is too new..

Keen to know if anyone has a working workaround solution.

**Comment by sigma67** [comment]
Fyi @ everyone here, Artifactory now has support for PyPI JSON API since 7.111

https://jfrog.com/help/r/jfrog-release-information/artifactory-7.111.12-self-hosted

**Comment by horgh** [comment]
I was able to get Renovate using timestamps from Artifactory by setting `registryUrls` to `"https://example.com/artifactory/api/pypi/pypi/pypi"` in Renovate's config. I kept my `pyproject.toml` index using the `/simple` URL, e.g. `https://example.com/artifactory/api/pypi/pypi/simple` in my case.

**Reply by horgh** [reply]
Doing that causes it to use the JSON endpoints. e.g. `https://example.com/artifactory/api/pypi/pypi/pypi/pytz/json`

**Reply by pascal-hofmann** [reply]
This does not work for me. Artifactories implementation of the [JSON API](https://docs.pypi.org/api/json/) does not return any release information under the `releases` key, and their support told me, that's because this is deprecated according to the spec.

The non-deprecated way to get release information is to use the [Index API](https://docs.pypi.org/api/index-api/), which is in fact the simple API. You can get the upload timestamp from this API, but you need to set `Content-type: application/vnd.pypi.simple.v1+json` to get the extended JSON response instead of the HTML response, see https://peps.python.org/pep-0691/.

Also, artifactory needs to be configured accordingly: https://docs.jfrog.com/artifactory/docs/pypi-repositories#enable-json-indexing-in-pypi-repositories

Edit: See https://github.com/renovatebot/renovate/pull/44222 for a solution to this. The config suggested above must be removed for this to work though. (And artifactory needs to be configured accordingly.)
